<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Perform full SCA analysis on the S1A serine protease dataset &#8212; cocoatree 0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=d7e2fa82" />
    <link rel="stylesheet" type="text/css" href="../_static/nameko.css?v=2296cfae" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css?v=d2d258e8" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery-binder.css?v=f4aeca0c" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery-dataframe.css?v=2082cf3c" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery-rendered-html.css?v=1277b6f3" />
    <script src="../_static/documentation_options.js?v=837179f8"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Lora:400' rel='stylesheet' type='text/css'>
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">

  </head><body><div class="navbar related">
      <a href="../index.html">
        
      </a>
    <ul class="nav nav-pills nav-stacked">
      <li class="right"><a href="index.html">Gallery</a> </li>
      <li class="right"><a href="../modules/classes.html">Documentation</a> | </li>
      <li class="right"><a href="../install.html">Installation</a> | </li>
      <li class="right"><a href="../index.html">Home</a> | </li>
      
    </ul>
  </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="sphx-glr-download-link-note admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#sphx-glr-download-auto-examples-plot-full-sca-analysis-py"><span class="std std-ref">Go to the end</span></a>
to download the full example code. or to run this example in your browser via Binder</p>
</div>
<section class="sphx-glr-example-title" id="perform-full-sca-analysis-on-the-s1a-serine-protease-dataset">
<span id="sphx-glr-auto-examples-plot-full-sca-analysis-py"></span><h1>Perform full SCA analysis on the S1A serine protease dataset<a class="headerlink" href="#perform-full-sca-analysis-on-the-s1a-serine-protease-dataset" title="Link to this heading">¶</a></h1>
<p>This example shows the full process to perform a complete coevolution
analysis in order to detect protein sectors from data importation, MSA
filtering, computation of positional and joint amino acid frequencies,
and computation of the SCA coevolution matrix.</p>
<p>The matrix is then decomposed into principal components and independent
component analysis is performed.</p>
<p>In the end, we export a fasta file of the residues contributing to the first
independent component.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Author: Margaux Jullien &lt;margaux.jullien@univ-grenoble-alpes.fr&gt;</span>
<span class="c1">#         Nelle Varoquaux &lt;nelle.varoquaux@univ-grenoble-alpes.fr&gt;</span>
<span class="c1">#         Ivan Junier &lt;ivan.junier@univ-grenoble-alpes.fr&gt;</span>
<span class="c1"># License: TBD</span>
</pre></div>
</div>
<p>Import necessary</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cocoatree.datasets</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_S1A_serine_proteases</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cocoatree.io</span><span class="w"> </span><span class="kn">import</span> <span class="n">export_fasta</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cocoatree.msa</span><span class="w"> </span><span class="kn">import</span> <span class="n">filter_gap_seq</span><span class="p">,</span> <span class="n">filter_gap_pos</span><span class="p">,</span> <span class="n">seq_weights</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cocoatree.statistics.position</span><span class="w"> </span><span class="kn">import</span> <span class="n">aa_freq_at_pos</span><span class="p">,</span> \
    <span class="n">compute_background_frequencies</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cocoatree.statistics.pairwise</span><span class="w"> </span><span class="kn">import</span> <span class="n">aa_joint_freq</span><span class="p">,</span> <span class="n">compute_sca_matrix</span><span class="p">,</span> \
    <span class="n">compute_seq_identity</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cocoatree.deconvolution</span><span class="w"> </span><span class="kn">import</span> <span class="n">eigen_decomp</span><span class="p">,</span> <span class="n">compute_ica</span><span class="p">,</span> \
    <span class="n">choose_num_components</span><span class="p">,</span> <span class="n">icList</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cocoatree.randomize</span><span class="w"> </span><span class="kn">import</span> <span class="n">randomization</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
</pre></div>
</div>
<section id="load-the-dataset">
<h2>Load the dataset<a class="headerlink" href="#load-the-dataset" title="Link to this heading">¶</a></h2>
<p>We start by importing the dataset. In this case, we can directly load the S1
serine protease dataset provided in <code class="xref py py-mod docutils literal notranslate"><span class="pre">cocoatree</span></code>. To work on your on
dataset, you can use the <a href="#id1"><span class="problematic" id="id2">:fun:`cocoatree.io.load_msa`</span></a> function.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">serine_dataset</span> <span class="o">=</span> <span class="n">load_S1A_serine_proteases</span><span class="p">()</span>
<span class="n">seq_id</span> <span class="o">=</span> <span class="n">serine_dataset</span><span class="p">[</span><span class="s2">&quot;sequence_ids&quot;</span><span class="p">]</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">serine_dataset</span><span class="p">[</span><span class="s2">&quot;alignment&quot;</span><span class="p">]</span>
<span class="n">n_pos</span><span class="p">,</span> <span class="n">n_seq</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The loaded MSA has </span><span class="si">{</span><span class="n">n_seq</span><span class="si">}</span><span class="s2"> sequences and </span><span class="si">{</span><span class="n">n_pos</span><span class="si">}</span><span class="s2"> positions.&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>The loaded MSA has 6970 sequences and 296 positions.
</pre></div>
</div>
</section>
<section id="msa-filtering">
<h2>MSA filtering<a class="headerlink" href="#msa-filtering" title="Link to this heading">¶</a></h2>
<p>We are going to clean a bit the loaded MSA by filtering both sequences and
positions.</p>
<section id="filter-overly-gapped-positions">
<h3>Filter overly gapped positions<a class="headerlink" href="#filter-overly-gapped-positions" title="Link to this heading">¶</a></h3>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">filt_seqs</span><span class="p">,</span> <span class="n">pos_kept</span> <span class="o">=</span> <span class="n">filter_gap_pos</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.4</span><span class="p">)</span>
<span class="n">n_pos_kept</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">pos_kept</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;After filtering, we have </span><span class="si">{</span><span class="n">n_pos_kept</span><span class="si">}</span><span class="s2"> remaining sequences.&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>After filtering, we have 226 remaining sequences.
</pre></div>
</div>
</section>
<section id="filter-overly-gapped-sequences">
<h3>Filter overly gapped sequences<a class="headerlink" href="#filter-overly-gapped-sequences" title="Link to this heading">¶</a></h3>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">seq_id_kept</span><span class="p">,</span> <span class="n">seq_kept</span> <span class="o">=</span> <span class="n">filter_gap_seq</span><span class="p">(</span><span class="n">seq_id</span><span class="p">,</span> <span class="n">filt_seqs</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
                                       <span class="n">filtrefseq</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="compute-the-matrix-of-pairwise-sequence-identity">
<h2>Compute the matrix of pairwise sequence identity<a class="headerlink" href="#compute-the-matrix-of-pairwise-sequence-identity" title="Link to this heading">¶</a></h2>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">sim_matrix</span> <span class="o">=</span> <span class="n">compute_seq_identity</span><span class="p">(</span><span class="n">seq_kept</span><span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">m</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">sim_matrix</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;inferno&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;sequences&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;sequences&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Matrix of pairwise sequence identity&#39;</span><span class="p">,</span> <span class="n">fontweight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
<span class="n">cb</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
<span class="n">cb</span><span class="o">.</span><span class="n">set_label</span><span class="p">(</span><span class="s2">&quot;Pairwise sequence identity&quot;</span><span class="p">,</span> <span class="n">fontweight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
</pre></div>
</div>
<img src="../_images/sphx_glr_plot_full_SCA_analysis_001.png" srcset="../_images/sphx_glr_plot_full_SCA_analysis_001.png" alt="Matrix of pairwise sequence identity" class = "sphx-glr-single-img"/><p>Compute sequence weights</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">weights</span><span class="p">,</span> <span class="n">n_eff_seq</span> <span class="o">=</span> <span class="n">seq_weights</span><span class="p">(</span><span class="n">sim_matrix</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of effective sequences </span><span class="si">{</span><span class="n">n_eff_seq</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>Number of effective sequences 1205.7209863932837
</pre></div>
</div>
<p>compute allele frequencies</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">aa_freq</span> <span class="o">=</span> <span class="n">aa_freq_at_pos</span><span class="p">(</span><span class="n">seq_kept</span><span class="p">,</span> <span class="n">lambda_coef</span><span class="o">=</span><span class="mf">0.03</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="n">weights</span><span class="p">)</span>
</pre></div>
</div>
<p>Compute background frequencies</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">background_frequencies</span> <span class="o">=</span> <span class="n">compute_background_frequencies</span><span class="p">(</span><span class="n">aa_freq</span><span class="p">)</span>
</pre></div>
</div>
<p>Compute joint allele frequencies</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">fijab</span><span class="p">,</span> <span class="n">fijab_ind</span> <span class="o">=</span> <span class="n">aa_joint_freq</span><span class="p">(</span><span class="n">seq_kept</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="n">weights</span><span class="p">,</span> <span class="n">lambda_coef</span><span class="o">=</span><span class="mf">0.03</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="compute-the-sca-coevolution-matrix">
<h2>Compute the SCA coevolution matrix<a class="headerlink" href="#compute-the-sca-coevolution-matrix" title="Link to this heading">¶</a></h2>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">Cijab_raw</span><span class="p">,</span> <span class="n">Cij</span> <span class="o">=</span> <span class="n">compute_sca_matrix</span><span class="p">(</span><span class="n">joint_freqs</span><span class="o">=</span><span class="n">fijab</span><span class="p">,</span>
                                    <span class="n">joint_freqs_ind</span><span class="o">=</span><span class="n">fijab_ind</span><span class="p">,</span>
                                    <span class="n">aa_freq</span><span class="o">=</span><span class="n">aa_freq</span><span class="p">,</span>
                                    <span class="n">background_freq</span><span class="o">=</span><span class="n">background_frequencies</span><span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">im</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">Cij</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mf">1.4</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;inferno&#39;</span><span class="p">)</span>

<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Coevolution matrix&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">im</span><span class="p">,</span> <span class="n">shrink</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
</pre></div>
</div>
<img src="../_images/sphx_glr_plot_full_SCA_analysis_002.png" srcset="../_images/sphx_glr_plot_full_SCA_analysis_002.png" alt="Coevolution matrix" class = "sphx-glr-single-img"/><div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>&lt;matplotlib.colorbar.Colorbar object at 0x7f2646bbb290&gt;
</pre></div>
</div>
<section id="decomposition-of-the-matrix-into-principal-components">
<h3>Decomposition of the matrix into principal components<a class="headerlink" href="#decomposition-of-the-matrix-into-principal-components" title="Link to this heading">¶</a></h3>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">eigenvalues</span><span class="p">,</span> <span class="n">eigenvectors</span> <span class="o">=</span> <span class="n">eigen_decomp</span><span class="p">(</span><span class="n">Cij</span><span class="p">)</span>
</pre></div>
</div>
<p>Plot distribution of eigenvalues</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">eigenvalues</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Number&#39;</span><span class="p">,</span> <span class="n">fontweight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Eigenvalue&#39;</span><span class="p">,</span> <span class="n">fontweight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
</pre></div>
</div>
<img src="../_images/sphx_glr_plot_full_SCA_analysis_003.png" srcset="../_images/sphx_glr_plot_full_SCA_analysis_003.png" alt="plot full SCA analysis" class = "sphx-glr-single-img"/><div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>Text(0.5, 23.52222222222222, &#39;Eigenvalue&#39;)
</pre></div>
</div>
</section>
<section id="select-number-of-significant-components">
<h3>Select number of significant components<a class="headerlink" href="#select-number-of-significant-components" title="Link to this heading">¶</a></h3>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">v_rand</span><span class="p">,</span> <span class="n">l_rand</span> <span class="o">=</span> <span class="n">randomization</span><span class="p">(</span><span class="n">seq_kept</span><span class="p">,</span> <span class="n">Nrep</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                               <span class="n">weights</span><span class="o">=</span><span class="n">weights</span><span class="p">,</span> <span class="n">lambda_coef</span><span class="o">=</span><span class="mf">0.03</span><span class="p">,</span> <span class="n">kmax</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                               <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;SCA&#39;</span><span class="p">,</span>
                               <span class="n">correction</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">n_components</span> <span class="o">=</span> <span class="n">choose_num_components</span><span class="p">(</span><span class="n">eigenvalues</span><span class="p">,</span> <span class="n">l_rand</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;n_components = &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">n_components</span><span class="p">))</span>

<span class="n">hist0</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="n">l_rand</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">bins</span><span class="o">=</span><span class="n">n_pos_kept</span><span class="p">,</span>
                           <span class="nb">range</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">eigenvalues</span><span class="o">.</span><span class="n">max</span><span class="p">()))</span>
<span class="n">hist1</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="n">eigenvalues</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">n_pos_kept</span><span class="p">,</span>
                           <span class="nb">range</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">eigenvalues</span><span class="o">.</span><span class="n">max</span><span class="p">()))</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">bins</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">hist1</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">bins</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">bins</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">hist0</span><span class="o">/</span><span class="mi">10</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Eigenvalues&#39;</span><span class="p">,</span> <span class="n">fontweight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Numbers&#39;</span><span class="p">,</span> <span class="n">fontweight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="sphx-glr-script-out highlight-pytb notranslate"><div class="highlight"><pre><span></span><span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;/home/runner/work/cocoatree/cocoatree/examples/plot_full_SCA_analysis.py&quot;</span>, line <span class="m">139</span>, in <span class="n">&lt;module&gt;</span>
<span class="w">    </span><span class="n">v_rand</span><span class="p">,</span> <span class="n">l_rand</span> <span class="o">=</span> <span class="n">randomization</span><span class="p">(</span><span class="n">seq_kept</span><span class="p">,</span> <span class="n">Nrep</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
<span class="w">                     </span><span class="pm">^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^</span>
<span class="gr">TypeError</span>: <span class="n">randomization() got an unexpected keyword argument &#39;Nrep&#39;</span>
</pre></div>
</div>
</section>
<section id="independent-component-analysis-ica">
<h3>Independent component analysis (ICA)<a class="headerlink" href="#independent-component-analysis-ica" title="Link to this heading">¶</a></h3>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">independant_components</span><span class="p">,</span> <span class="n">W</span> <span class="o">=</span> <span class="n">compute_ica</span><span class="p">(</span>
    <span class="n">eigenvectors</span><span class="p">,</span> <span class="n">kmax</span><span class="o">=</span><span class="n">n_components</span><span class="p">,</span> <span class="n">learnrate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">iterations</span><span class="o">=</span><span class="mi">100000</span><span class="p">)</span>

<span class="c1"># Plot results</span>
<span class="k">if</span> <span class="n">n_components</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Uneven number of axes, discard the last one for visual </span><span class="se">\</span>
<span class="s1">          representation&#39;</span><span class="p">)</span>
    <span class="n">n_components</span> <span class="o">-=</span> <span class="mi">2</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">n_components</span> <span class="o">-=</span> <span class="mi">1</span>

<span class="n">pairs</span> <span class="o">=</span> <span class="p">[[</span><span class="n">x</span><span class="p">,</span> <span class="n">x</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">n_components</span><span class="p">,</span> <span class="mi">2</span><span class="p">)]</span>
<span class="n">ncols</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">pairs</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.figsize&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">14</span><span class="p">,</span> <span class="mi">8</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">pairs</span><span class="p">),</span> <span class="n">tight_layout</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="p">[</span><span class="n">k1</span><span class="p">,</span> <span class="n">k2</span><span class="p">]</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pairs</span><span class="p">):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">k</span><span class="p">]</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">eigenvectors</span><span class="p">[:,</span> <span class="n">k1</span><span class="p">],</span> <span class="n">eigenvectors</span><span class="p">[:,</span> <span class="n">k2</span><span class="p">],</span> <span class="s1">&#39;ok&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;eigenvector </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">k1</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;eigenvector </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">k2</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>

    <span class="n">ax</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="n">k</span><span class="p">]</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">independant_components</span><span class="p">[:,</span> <span class="n">k1</span><span class="p">],</span> <span class="n">independant_components</span><span class="p">[:,</span> <span class="n">k2</span><span class="p">],</span> <span class="s1">&#39;ok&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;independant component </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">k1</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;independant component </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">k2</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
</pre></div>
</div>
<p>Select residues that significantly contribute to each independent component</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">ics</span><span class="p">,</span> <span class="n">icsize</span><span class="p">,</span> <span class="n">sortedpos</span><span class="p">,</span> <span class="n">cutoff</span><span class="p">,</span> <span class="n">scaled_pdf</span><span class="p">,</span> <span class="n">all_fits</span> <span class="o">=</span> <span class="n">icList</span><span class="p">(</span>
    <span class="n">independant_components</span><span class="p">,</span> <span class="n">n_components</span><span class="p">,</span> <span class="n">Cij</span><span class="p">,</span>
    <span class="n">p_cut</span><span class="o">=</span><span class="mf">0.95</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Sizes of the </span><span class="si">{</span><span class="n">n_components</span><span class="si">}</span><span class="s2"> ICs: </span><span class="si">{</span><span class="n">icsize</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Plot coevolution within and between the sectors</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">tight_layout</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">im</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">Cij</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ix_</span><span class="p">(</span><span class="n">sortedpos</span><span class="p">,</span> <span class="n">sortedpos</span><span class="p">)],</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
               <span class="n">interpolation</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span>
               <span class="n">extent</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">icsize</span><span class="p">),</span> <span class="mi">0</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">icsize</span><span class="p">)],</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;inferno&#39;</span><span class="p">)</span>
<span class="n">cb</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">im</span><span class="p">)</span>
<span class="n">cb</span><span class="o">.</span><span class="n">set_label</span><span class="p">(</span><span class="s2">&quot;Coevolution measure&quot;</span><span class="p">)</span>

<span class="n">line_index</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_components</span><span class="p">):</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">line_index</span> <span class="o">+</span> <span class="n">icsize</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">line_index</span> <span class="o">+</span> <span class="n">icsize</span><span class="p">[</span><span class="n">i</span><span class="p">]],</span>
            <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">icsize</span><span class="p">)],</span> <span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">icsize</span><span class="p">)],</span> <span class="p">[</span><span class="nb">sum</span><span class="p">(</span><span class="n">icsize</span><span class="p">)</span> <span class="o">-</span> <span class="n">line_index</span><span class="p">,</span>
                               <span class="nb">sum</span><span class="p">(</span><span class="n">icsize</span><span class="p">)</span> <span class="o">-</span> <span class="n">line_index</span><span class="p">],</span> <span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">line_index</span> <span class="o">+=</span> <span class="n">icsize</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
</pre></div>
</div>
<p>Export fasta files of the sectors for all the sequences</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">sector_1_pos</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pos_kept</span><span class="p">[</span><span class="n">ics</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">])</span>
<span class="n">sector_1</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq_id</span><span class="p">)):</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">sector_1_pos</span><span class="p">:</span>
        <span class="n">seq</span> <span class="o">+=</span> <span class="n">sequences</span><span class="p">[</span><span class="n">sequence</span><span class="p">][</span><span class="n">pos</span><span class="p">]</span>
    <span class="n">sector_1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

<span class="n">export_fasta</span><span class="p">(</span><span class="n">sector_1</span><span class="p">,</span> <span class="n">seq_id</span><span class="p">,</span> <span class="s1">&#39;sector_1.fasta&#39;</span><span class="p">)</span>
<span class="c1"># outpath needs to be defined properly</span>
</pre></div>
</div>
<p class="sphx-glr-timing"><strong>Total running time of the script:</strong> (0 minutes 13.475 seconds)</p>
<div class="sphx-glr-footer sphx-glr-footer-example docutils container" id="sphx-glr-download-auto-examples-plot-full-sca-analysis-py">
<div class="binder-badge docutils container">
<a class="reference external image-reference" href="https://mybinder.org/v2/gh/tree-timc/cocoatree/gh-pages?urlpath=lab/tree/notebooks/auto_examples/plot_full_SCA_analysis.ipynb"><img alt="Launch binder" src="../_images/binder_badge_logo1.svg" style="width: 150px;" />
</a>
</div>
<div class="sphx-glr-download sphx-glr-download-jupyter docutils container">
<p><a class="reference download internal" download="" href="../_downloads/1e714e784dd866ce7d4944d2a68a69c1/plot_full_SCA_analysis.ipynb"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">Jupyter</span> <span class="pre">notebook:</span> <span class="pre">plot_full_SCA_analysis.ipynb</span></code></a></p>
</div>
<div class="sphx-glr-download sphx-glr-download-python docutils container">
<p><a class="reference download internal" download="" href="../_downloads/d1f0552ffc7b30b60c5aa88d0e9443c7/plot_full_SCA_analysis.py"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">Python</span> <span class="pre">source</span> <span class="pre">code:</span> <span class="pre">plot_full_SCA_analysis.py</span></code></a></p>
</div>
<div class="sphx-glr-download sphx-glr-download-zip docutils container">
<p><a class="reference download internal" download="" href="../_downloads/45b16a9623f4fa80310e2189f76b5f1b/plot_full_SCA_analysis.zip"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">zipped:</span> <span class="pre">plot_full_SCA_analysis.zip</span></code></a></p>
</div>
</div>
<p class="sphx-glr-signature"><a class="reference external" href="https://sphinx-gallery.github.io">Gallery generated by Sphinx-Gallery</a></p>
</section>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
  </body>
</html>