<!DOCTYPE html>

<html lang="en" data-content_root="../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>cocoatree.io &#8212; cocoatree 0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=d7e2fa82" />
    <link rel="stylesheet" type="text/css" href="../../_static/nameko.css?v=2296cfae" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery.css?v=d2d258e8" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-binder.css?v=f4aeca0c" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-dataframe.css?v=2082cf3c" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-rendered-html.css?v=1277b6f3" />
    <script src="../../_static/documentation_options.js?v=837179f8"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Lora:400' rel='stylesheet' type='text/css'>
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">

  </head><body><div class="navbar related">
      <a href="../../index.html">
        
      </a>
    <ul class="nav nav-pills nav-stacked">
      <li class="right"><a href="../../auto_examples/index.html">Gallery</a> </li>
      <li class="right"><a href="../../modules/classes.html">API</a> | </li>
      <li class="right"><a href="../../install.html">Installation</a> | </li>
      <li class="right"><a href="../../index.html">Home</a> | </li>
      
    </ul>
  </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for cocoatree.io</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">Bio</span><span class="w"> </span><span class="kn">import</span> <span class="n">AlignIO</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB</span><span class="w"> </span><span class="kn">import</span> <span class="n">PDBParser</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">.msa</span><span class="w"> </span><span class="kn">import</span> <span class="n">_clean_msa</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">.__params</span><span class="w"> </span><span class="kn">import</span> <span class="n">aatable</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">ete3</span><span class="w"> </span><span class="kn">import</span> <span class="n">Tree</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>


<div class="viewcode-block" id="load_MSA">
<a class="viewcode-back" href="../../modules/generated/cocoatree.io.load_MSA.html#cocoatree.io.load_MSA">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">load_MSA</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">,</span> <span class="n">clean</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Read in a multiple sequence alignment (MSA)</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file_path : path to the alignment file</span>

<span class="sd">    format : string {&quot;fasta&quot;, &quot;phylip&quot;, â€¦}, optional, default: &quot;fasta&quot;</span>
<span class="sd">        format of the alignment file (e.g. &#39;fasta&#39;, &#39;phylip&#39;, etc.)</span>
<span class="sd">        All format supported by biopython&#39;s Bio.AlignIO.read are accepted.</span>

<span class="sd">    clean : boolean, default=True</span>
<span class="sd">        whether to remove ambiguous amino acids (e.g. B, X etc.)</span>

<span class="sd">    verbose : boolean,</span>
<span class="sd">            whether to print informations about the MSA</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    a dictionnary containing:</span>
<span class="sd">        - `sequences_id`, list of sequence identifiers</span>
<span class="sd">        - `alignment`: list of sequences as strings</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">alignment</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="nb">format</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">clean</span><span class="p">:</span>
        <span class="n">alignment</span> <span class="o">=</span> <span class="n">_clean_msa</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>

    <span class="n">sequences_id</span> <span class="o">=</span> <span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">alignment</span><span class="p">]</span>
    <span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">alignment</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Number of sequences: </span><span class="si">%i</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Alignment of length: </span><span class="si">%i</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>

    <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;sequence_ids&quot;</span><span class="p">:</span> <span class="n">sequences_id</span><span class="p">,</span> <span class="s2">&quot;alignment&quot;</span><span class="p">:</span> <span class="n">sequences</span><span class="p">}</span></div>



<div class="viewcode-block" id="load_tree_ete3">
<a class="viewcode-back" href="../../modules/generated/cocoatree.io.load_tree_ete3.html#cocoatree.io.load_tree_ete3">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">load_tree_ete3</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    From the loading of a Newick tree, generate a ete3.Tree object</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file_path : path to the Newick file</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    tree_ete3 : `ete3.Tree` object</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tree_ete3</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">tree_ete3</span></div>



<div class="viewcode-block" id="export_fasta">
<a class="viewcode-back" href="../../modules/generated/cocoatree.io.export_fasta.html#cocoatree.io.export_fasta">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">export_fasta</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">sequences_id</span><span class="p">,</span> <span class="n">outpath</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Export intermediate files in FASTA format</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    sequences : list of sequences as strings (as imported by load_MSA)</span>

<span class="sd">    sequences_id : list of sequences identifiers (as imported by load_MSA)</span>

<span class="sd">    outpath : path to the output file</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Add checks to see if the path exists?</span>
    <span class="n">Nseq</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">Nseq</span><span class="p">):</span>
            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequences_id</span><span class="p">[</span><span class="n">record</span><span class="p">])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">sequences</span><span class="p">[</span><span class="n">record</span><span class="p">])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span></div>



<div class="viewcode-block" id="load_pdb">
<a class="viewcode-back" href="../../modules/generated/cocoatree.io.load_pdb.html#cocoatree.io.load_pdb">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">load_pdb</span><span class="p">(</span><span class="n">path2pdb</span><span class="p">,</span> <span class="n">pdb_id</span><span class="p">,</span> <span class="n">chain</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read in a PDB file.</span>

<span class="sd">    Import a PDB file and extract the associated sequence along with the</span>
<span class="sd">    amino acid positions</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    path2pdb : path to the PDB file</span>

<span class="sd">    pdb_id : str,</span>
<span class="sd">        identifier of the PDB file</span>

<span class="sd">    chain : str,</span>
<span class="sd">        name of the chain to read</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pbd_seq : str,</span>
<span class="sd">        amino acid sequence of the PDB file</span>

<span class="sd">    pdb_pos : list,</span>
<span class="sd">        PDB position of each amino acid</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">warnings</span><span class="o">.</span><span class="n">catch_warnings</span><span class="p">():</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
        <span class="n">P</span> <span class="o">=</span> <span class="n">PDBParser</span><span class="p">(</span><span class="n">PERMISSIVE</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">structure</span> <span class="o">=</span> <span class="n">P</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">,</span> <span class="n">path2pdb</span><span class="p">)</span>

    <span class="c1"># Fill up sequence and label information</span>
    <span class="n">pdb_seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">pdb_pos</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="n">residues</span> <span class="o">=</span> <span class="p">[</span><span class="n">res</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">structure</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">chain</span><span class="p">]</span> <span class="k">if</span> <span class="n">res</span><span class="o">.</span><span class="n">get_id</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot; &quot;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">residues</span><span class="p">:</span>
        <span class="n">pdb_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">get_id</span><span class="p">()[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">get_id</span><span class="p">()[</span><span class="mi">2</span><span class="p">])</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">pdb_seq</span> <span class="o">+=</span> <span class="n">aatable</span><span class="p">[</span><span class="n">res</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()]</span>
        <span class="k">except</span> <span class="ne">BaseException</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Error: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
            <span class="n">pdb_seq</span> <span class="o">+=</span> <span class="s2">&quot;X&quot;</span>

    <span class="k">return</span> <span class="n">pdb_seq</span><span class="p">,</span> <span class="n">pdb_pos</span></div>



<div class="viewcode-block" id="export_sector_for_pymol">
<a class="viewcode-back" href="../../modules/generated/cocoatree.io.export_sector_for_pymol.html#cocoatree.io.export_sector_for_pymol">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">export_sector_for_pymol</span><span class="p">(</span><span class="n">mapping</span><span class="p">,</span> <span class="n">independent_components</span><span class="p">,</span> <span class="n">axis</span><span class="p">,</span>
                            <span class="n">sector_pos_in_loaded_msa</span><span class="p">,</span>
                            <span class="n">sector_pos_in_filtered_msa</span><span class="p">,</span>
                            <span class="n">outpath</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Export sector information for mapping on 3D structure in PyMOL.</span>

<span class="sd">    Export numpy arrays of a sector&#39;s residue positions and their contribution</span>
<span class="sd">    for coloring in PyMOL.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    mapping : numpy.ndarray,</span>
<span class="sd">        mapping between the unfiltered MSA and the PDB structure, output of</span>
<span class="sd">        cocoatree.msa.map_to_pdb() function</span>

<span class="sd">    independent_components : numpy.ndarray,</span>
<span class="sd">        output of cocoatree.deconvolution.compute_ica() function</span>

<span class="sd">    axis : int,</span>
<span class="sd">        rank of the independent component associated with the desired sector</span>

<span class="sd">    sector_pos_in_loaded_msa : list,</span>
<span class="sd">        positions of the sector&#39;s residues in the unfiltered MSA</span>

<span class="sd">    sector_pos_in_filtered_msa : numpy.ndarray,</span>
<span class="sd">        positions of the sector&#39;s residues in the filtered MSA, output from</span>
<span class="sd">        cocoatree.deconvolution.icList() function</span>

<span class="sd">    outpath : str,</span>
<span class="sd">        path to the output file as a binary in .npy format</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    binary file in .npy format containing an array with the positions of the</span>
<span class="sd">    sector&#39;s residues and an array with their contribution to the independent</span>
<span class="sd">    component.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">sector_pdb_pos</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">sector_pos_in_loaded_msa</span><span class="p">:</span>
        <span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mapping</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">residue</span><span class="p">))[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">sector_pdb_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mapping</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">index</span><span class="p">])</span>

    <span class="n">ic_contributions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">sector_pos_in_filtered_msa</span><span class="p">:</span>
        <span class="n">ic_contributions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">independent_components</span><span class="p">[</span><span class="n">residue</span><span class="p">,</span> <span class="n">axis</span><span class="p">])</span>

    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">sector_pdb_pos</span><span class="p">,</span> <span class="n">ic_contributions</span><span class="p">]))</span></div>

</pre></div>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
  </body>
</html>