<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Reference &#8212; cocoatree 0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=d7e2fa82" />
    <link rel="stylesheet" type="text/css" href="../_static/nameko.css?v=2296cfae" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css?v=d2d258e8" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery-binder.css?v=f4aeca0c" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery-dataframe.css?v=2082cf3c" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery-rendered-html.css?v=1277b6f3" />
    <script src="../_static/documentation_options.js?v=837179f8"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Lora:400' rel='stylesheet' type='text/css'>
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">

  </head><body><div class="navbar related">
      <a href="../index.html">
        
      </a>
    <ul class="nav nav-pills nav-stacked">
      <li class="right"><a href="../auto_examples/index.html">Gallery</a> </li>
      <li class="right"><a href="#">Documentation</a> | </li>
      <li class="right"><a href="../install.html">Installation</a> | </li>
      <li class="right"><a href="../index.html">Home</a> | </li>
      
    </ul>
  </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="reference">
<h1>Reference<a class="headerlink" href="#reference" title="Link to this heading">¶</a></h1>
<section id="module-cocoatree.io">
<span id="cocoatree-io-functions-to-import-and-export-files"></span><h2><a class="reference internal" href="#module-cocoatree.io" title="cocoatree.io"><code class="xref py py-mod docutils literal notranslate"><span class="pre">cocoatree.io</span></code></a>: Functions to import and export files<a class="headerlink" href="#module-cocoatree.io" title="Link to this heading">¶</a></h2>
<section id="functions">
<h3>Functions<a class="headerlink" href="#functions" title="Link to this heading">¶</a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.io.load_MSA.html#cocoatree.io.load_MSA" title="cocoatree.io.load_MSA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">io.load_MSA</span></code></a>(file_path, format[, clean, verbose])</p></td>
<td><p>Read in a multiple sequence alignment (MSA)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/cocoatree.io.load_tree_ete3.html#cocoatree.io.load_tree_ete3" title="cocoatree.io.load_tree_ete3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">io.load_tree_ete3</span></code></a>(file_path)</p></td>
<td><p>From the loading of a Newick tree, generate a ete3.Tree object</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.io.export_fasta.html#cocoatree.io.export_fasta" title="cocoatree.io.export_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">io.export_fasta</span></code></a>(sequences, seq_id, outpath)</p></td>
<td><p>Function to export intermediate files in fasta format</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/cocoatree.io.load_pdb.html#cocoatree.io.load_pdb" title="cocoatree.io.load_pdb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">io.load_pdb</span></code></a>(path2pdb, pdb_id, chain)</p></td>
<td><p>Import a PDB file and extract the associated sequence along with the amino acid positions</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.io.export_sector_for_pymol.html#cocoatree.io.export_sector_for_pymol" title="cocoatree.io.export_sector_for_pymol"><code class="xref py py-obj docutils literal notranslate"><span class="pre">io.export_sector_for_pymol</span></code></a>(mapping, ...)</p></td>
<td><p>Export numpy arrays of a sector's residue positions and their contribution for coloring in PyMol.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="module-cocoatree.datasets">
<span id="cocoatree-datasets-datasets"></span><span id="io-ref"></span><h2><a class="reference internal" href="#module-cocoatree.datasets" title="cocoatree.datasets"><code class="xref py py-mod docutils literal notranslate"><span class="pre">cocoatree.datasets</span></code></a>: Datasets<a class="headerlink" href="#module-cocoatree.datasets" title="Link to this heading">¶</a></h2>
<section id="id1">
<h3>Functions<a class="headerlink" href="#id1" title="Link to this heading">¶</a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.datasets.load_S1A_serine_proteases.html#cocoatree.datasets.load_S1A_serine_proteases" title="cocoatree.datasets.load_S1A_serine_proteases"><code class="xref py py-obj docutils literal notranslate"><span class="pre">datasets.load_S1A_serine_proteases</span></code></a>()</p></td>
<td><p>Load the S1A serine protease dataset</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/cocoatree.datasets.load_rhomboid_proteases.html#cocoatree.datasets.load_rhomboid_proteases" title="cocoatree.datasets.load_rhomboid_proteases"><code class="xref py py-obj docutils literal notranslate"><span class="pre">datasets.load_rhomboid_proteases</span></code></a>()</p></td>
<td><p>Load the rhomboid protease dataset</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="module-cocoatree.msa">
<span id="cocoatree-msa-multiple-sequence-alignment"></span><span id="datasets-ref"></span><h2><a class="reference internal" href="#module-cocoatree.msa" title="cocoatree.msa"><code class="xref py py-mod docutils literal notranslate"><span class="pre">cocoatree.msa</span></code></a>: Multiple sequence alignment<a class="headerlink" href="#module-cocoatree.msa" title="Link to this heading">¶</a></h2>
<section id="id2">
<h3>Functions<a class="headerlink" href="#id2" title="Link to this heading">¶</a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.msa.filter_sequences.html#cocoatree.msa.filter_sequences" title="cocoatree.msa.filter_sequences"><code class="xref py py-obj docutils literal notranslate"><span class="pre">msa.filter_sequences</span></code></a>(seq_id, sequences[, ...])</p></td>
<td><p>Filter sequences</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/cocoatree.msa.filter_seq_id.html#cocoatree.msa.filter_seq_id" title="cocoatree.msa.filter_seq_id"><code class="xref py py-obj docutils literal notranslate"><span class="pre">msa.filter_seq_id</span></code></a>(seq_id, sequences, list_id)</p></td>
<td><p>Filter sequences based on list</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.msa.choose_ref_seq.html#cocoatree.msa.choose_ref_seq" title="cocoatree.msa.choose_ref_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">msa.choose_ref_seq</span></code></a>(msa)</p></td>
<td><p>This function chooses a default reference sequence for the alignment by taking the sequence which has the mean pairwise sequence identity closest to that of the entire sequence alignment.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/cocoatree.msa.map_to_pdb.html#cocoatree.msa.map_to_pdb" title="cocoatree.msa.map_to_pdb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">msa.map_to_pdb</span></code></a>(pdb_seq, pdb_pos, sequences, ...)</p></td>
<td><p>Mapping of the unfiltered MSA positions on a PDB structure.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.msa.compute_seq_identity.html#cocoatree.msa.compute_seq_identity" title="cocoatree.msa.compute_seq_identity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">msa.compute_seq_identity</span></code></a>(sequences)</p></td>
<td><p>Computes the identity between sequences in a MSA (as Hamming's pairwise distance)</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="module-cocoatree.statistics">
<span id="cocoatree-statistics-computation-of-statistics"></span><span id="msa-ref"></span><h2><a class="reference internal" href="#module-cocoatree.statistics" title="cocoatree.statistics"><code class="xref py py-mod docutils literal notranslate"><span class="pre">cocoatree.statistics</span></code></a>: Computation of statistics<a class="headerlink" href="#module-cocoatree.statistics" title="Link to this heading">¶</a></h2>
<section id="id3">
<h3>Functions<a class="headerlink" href="#id3" title="Link to this heading">¶</a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.statistics.compute_frequencies.html#cocoatree.statistics.compute_frequencies" title="cocoatree.statistics.compute_frequencies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">statistics.compute_frequencies</span></code></a>(sequences[, ...])</p></td>
<td><p>Compute frequencies on sequences</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/cocoatree.statistics.pairwise.compute_sca_matrix.html#cocoatree.statistics.pairwise.compute_sca_matrix" title="cocoatree.statistics.pairwise.compute_sca_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">statistics.pairwise.compute_sca_matrix</span></code></a>(...)</p></td>
<td><p>Compute the SCA coevolution matrix</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.statistics.pairwise.compute_mutual_information_matrix.html#cocoatree.statistics.pairwise.compute_mutual_information_matrix" title="cocoatree.statistics.pairwise.compute_mutual_information_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">statistics.pairwise.compute_mutual_information_matrix</span></code></a>(...)</p></td>
<td><p>Compute the mutual information matrix</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/cocoatree.statistics.pairwise.compute_apc.html#cocoatree.statistics.pairwise.compute_apc" title="cocoatree.statistics.pairwise.compute_apc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">statistics.pairwise.compute_apc</span></code></a>(MIij)</p></td>
<td><p>Computes the average product correction (APC) as described in Dunn et al. (2008).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.statistics.pairwise.compute_entropy_correction.html#cocoatree.statistics.pairwise.compute_entropy_correction" title="cocoatree.statistics.pairwise.compute_entropy_correction"><code class="xref py py-obj docutils literal notranslate"><span class="pre">statistics.pairwise.compute_entropy_correction</span></code></a>(...)</p></td>
<td><p>Computes the entropy correction according to Vorberg et al. (2018).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/cocoatree.statistics.position.compute_entropy.html#cocoatree.statistics.position.compute_entropy" title="cocoatree.statistics.position.compute_entropy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">statistics.position.compute_entropy</span></code></a>(aa_freq)</p></td>
<td><p>Computes  Shannon's entropy for each position in the alignment</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.statistics.position.compute_rel_entropy.html#cocoatree.statistics.position.compute_rel_entropy" title="cocoatree.statistics.position.compute_rel_entropy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">statistics.position.compute_rel_entropy</span></code></a>(...)</p></td>
<td><p>Compute the relative entropy</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="module-cocoatree.deconvolution">
<span id="cocoatree-deconvolution-matrix-deconvolution"></span><span id="statistics-ref"></span><h2><a class="reference internal" href="#module-cocoatree.deconvolution" title="cocoatree.deconvolution"><code class="xref py py-mod docutils literal notranslate"><span class="pre">cocoatree.deconvolution</span></code></a>: Matrix deconvolution<a class="headerlink" href="#module-cocoatree.deconvolution" title="Link to this heading">¶</a></h2>
<section id="id4">
<h3>Functions<a class="headerlink" href="#id4" title="Link to this heading">¶</a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.deconvolution.eigen_decomp.html#cocoatree.deconvolution.eigen_decomp" title="cocoatree.deconvolution.eigen_decomp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">deconvolution.eigen_decomp</span></code></a>(mat)</p></td>
<td><p>Performs eigenvalue decomposition of a square matrix</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/cocoatree.deconvolution.compute_ica.html#cocoatree.deconvolution.compute_ica" title="cocoatree.deconvolution.compute_ica"><code class="xref py py-obj docutils literal notranslate"><span class="pre">deconvolution.compute_ica</span></code></a>(V[, kmax, ...])</p></td>
<td><p>ICA rotation (using _basicICA) with default parameters and normalization of outputs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.deconvolution.choose_num_components.html#cocoatree.deconvolution.choose_num_components" title="cocoatree.deconvolution.choose_num_components"><code class="xref py py-obj docutils literal notranslate"><span class="pre">deconvolution.choose_num_components</span></code></a>(...)</p></td>
<td><p>Given the eigenvalues of the coevolution matrix (eigenvalues), and the eigenvalues for the set of randomized matrices (rand_eigenvalues), return the number of significant eigenmodes as those above the average first eigenvalue plus 3 standard deviations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/cocoatree.deconvolution.extract_positions_from_IC.html#cocoatree.deconvolution.extract_positions_from_IC" title="cocoatree.deconvolution.extract_positions_from_IC"><code class="xref py py-obj docutils literal notranslate"><span class="pre">deconvolution.extract_positions_from_IC</span></code></a>(...)</p></td>
<td><p>Produces a list of positions contributing to each independent component (IC) above a defined statistical cutoff (p_cut, the cutoff on the CDF of the t-distribution fit to the histogram of each IC).</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="module-cocoatree.randomize">
<span id="cocoatree-randomize-alignment-randomization"></span><span id="deconvolution-ref"></span><h2><a class="reference internal" href="#module-cocoatree.randomize" title="cocoatree.randomize"><code class="xref py py-mod docutils literal notranslate"><span class="pre">cocoatree.randomize</span></code></a>: Alignment randomization<a class="headerlink" href="#module-cocoatree.randomize" title="Link to this heading">¶</a></h2>
<p>Module to perform randomization of alignments</p>
<section id="id5">
<h3>Functions<a class="headerlink" href="#id5" title="Link to this heading">¶</a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.deconvolution.eigen_decomp.html#cocoatree.deconvolution.eigen_decomp" title="cocoatree.deconvolution.eigen_decomp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">deconvolution.eigen_decomp</span></code></a>(mat)</p></td>
<td><p>Performs eigenvalue decomposition of a square matrix</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/cocoatree.deconvolution.compute_ica.html#cocoatree.deconvolution.compute_ica" title="cocoatree.deconvolution.compute_ica"><code class="xref py py-obj docutils literal notranslate"><span class="pre">deconvolution.compute_ica</span></code></a>(V[, kmax, ...])</p></td>
<td><p>ICA rotation (using _basicICA) with default parameters and normalization of outputs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.randomize.randomization.html#cocoatree.randomize.randomization" title="cocoatree.randomize.randomization"><code class="xref py py-obj docutils literal notranslate"><span class="pre">randomize.randomization</span></code></a>(sequences, n_rep[, ...])</p></td>
<td><p>Randomize the alignment while preserving the frequencies of amino acids at each position and compute the resulting spectrum of coevolution matrix.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="module-cocoatree.visualization">
<span id="cocoatree-visualization-visualization"></span><span id="randomize-ref"></span><h2><a class="reference internal" href="#module-cocoatree.visualization" title="cocoatree.visualization"><code class="xref py py-mod docutils literal notranslate"><span class="pre">cocoatree.visualization</span></code></a>: Visualization<a class="headerlink" href="#module-cocoatree.visualization" title="Link to this heading">¶</a></h2>
<p>Module to visualize phylogenetic trees along with sectors</p>
<section id="id6">
<h3>Functions<a class="headerlink" href="#id6" title="Link to this heading">¶</a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/cocoatree.visualization.update_tree_ete3_and_return_style.html#cocoatree.visualization.update_tree_ete3_and_return_style" title="cocoatree.visualization.update_tree_ete3_and_return_style"><code class="xref py py-obj docutils literal notranslate"><span class="pre">visualization.update_tree_ete3_and_return_style</span></code></a>(...)</p></td>
<td><p>Update ete3 tree with sector info and attributes and return tree_style for further visualization.</p></td>
</tr>
</tbody>
</table>
<span class="target" id="visualization-ref"></span></section>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
  </body>
</html>