<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>COCOA-Tree: COllaborative COevolution Analysis Toolbox including tree-based vizualisation &#8212; cocoatree 0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=d7e2fa82" />
    <link rel="stylesheet" type="text/css" href="_static/nameko.css?v=2296cfae" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery.css?v=d2d258e8" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery-binder.css?v=f4aeca0c" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery-dataframe.css?v=2082cf3c" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery-rendered-html.css?v=1277b6f3" />
    <script src="_static/documentation_options.js?v=837179f8"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Color sector residues on 3D structure in Pymol" href="interface_cocoatree_pymol.html" />
   
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Lora:400' rel='stylesheet' type='text/css'>
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">

  </head><body><div class="navbar related">
      <a href="#">
        
      </a>
    <ul class="nav nav-pills nav-stacked">
      <li class="right"><a href="auto_examples/index.html">Gallery</a> </li>
      <li class="right"><a href="modules/classes.html">API</a> | </li>
      <li class="right"><a href="install.html">Installation</a> | </li>
      <li class="right"><a href="#">Home</a> | </li>
      
    </ul>
  </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="toctree-wrapper compound">
</div>
<section id="cocoa-tree-collaborative-coevolution-analysis-toolbox-including-tree-based-vizualisation">
<h1>COCOA-Tree: COllaborative COevolution Analysis Toolbox including tree-based vizualisation<a class="headerlink" href="#cocoa-tree-collaborative-coevolution-analysis-toolbox-including-tree-based-vizualisation" title="Link to this heading">¶</a></h1>
<p>COCOA-Tree is a <a class="reference external" href="https://www.python.org">Python</a> library designed to perform coevolution analysis of protein sequence data and to compare the results with phylogenetic trees and metadata. By doing so, it facilitates the investigation of the origins of coevolution patterns, their relationships to phylogeny and functional protein annotation, and ultimately aims to democratize the use and development of so-called <em>protein sectors</em> <span id="id1">[<a class="reference internal" href="#id7" title="Olivier Rivoire, Kimberly A. Reynolds, and Rama Ranganathan. Evolution-Based Functional Decomposition of Proteins. PLOS Computational Biology, 12(6):e1004817, June 2016. doi:10.1371/journal.pcbi.1004817.">RRR16</a>]</span> and <em>specificity-determining positions (SDPs)</em> <span id="id2">[<a class="reference internal" href="#id8" title="David De Juan, Florencio Pazos, and Alfonso Valencia. Emerging methods in protein co-evolution. Nature Reviews Genetics, 14(4):249–261, April 2013. doi:10.1038/nrg3414.">DJPV13</a>]</span>.</p>
<p>The library includes several coevolution metrics (and allows users to define their own), along with various correction methods, enabling easy comparison between approaches. It also integrates with the molecular visualization software <a class="reference external" href="https://pymol.org">PyMOL</a>, allowing users to map results onto 3D structures, more specifically, extremal co-evolving residues that form the basis of protein sectors <span id="id3">[<a class="reference internal" href="#id9" title="Margaux et al. Tbd. BioRxiv, 1(1):1, 2025. doi:no DOI yet.">eal25</a>]</span>.</p>
<p>The software is organized into different modules, detailed below:</p>
<a class="reference internal image-reference" href="schemas/draft_fig1_v3.png"><img alt="schemas/draft_fig1_v3.png" src="schemas/draft_fig1_v3.png" style="width: 100%;" />
</a>
<section id="list-of-abbreviations">
<h2>List of abbreviations<a class="headerlink" href="#list-of-abbreviations" title="Link to this heading">¶</a></h2>
<ul class="simple">
<li><p>MI: Mutual Information</p></li>
<li><p>MSA: Multiple Sequence Alignment</p></li>
<li><p>SCA: Statistical Coupling Analysis, the method underlying the identification of protein sectors <span id="id4">[<a class="reference internal" href="#id6" title="Najeeb Halabi, Olivier Rivoire, Stanislas Leibler, and Rama Ranganathan. Protein Sectors: Evolutionary Units of Three-Dimensional Structure. Cell, 138(4):774–786, August 2009. URL: https://linkinghub.elsevier.com/retrieve/pii/S0092867409009635 (visited on 2023-07-19), doi:10.1016/j.cell.2009.07.038.">HRLR09</a>, <a class="reference internal" href="#id7" title="Olivier Rivoire, Kimberly A. Reynolds, and Rama Ranganathan. Evolution-Based Functional Decomposition of Proteins. PLOS Computational Biology, 12(6):e1004817, June 2016. doi:10.1371/journal.pcbi.1004817.">RRR16</a>]</span></p></li>
<li><p>XCoR: eXtremal Co-evolving Residues</p></li>
</ul>
</section>
<section id="example-gallery">
<h2>Example gallery<a class="headerlink" href="#example-gallery" title="Link to this heading">¶</a></h2>
<p>Various examples of COCOA-Tree’s use can be found in the <a class="reference external" href="auto_examples/index.html">Gallery</a>.</p>
<p>If you wish to perform a minimal SCA analysis, go to
<a class="reference external" href="../examples/a_quick_start/plot_minimal_sca.py">Minimal SCA example</a>.</p>
<p>For a detailed SCA analysis, go to
<a class="reference external" href="../examples/b_advanced/plot_full_SCA_analysis.py">Full SCA analysis</a>.</p>
</section>
<section id="citing-cocoa-tree">
<h2>Citing COCOA-Tree<a class="headerlink" href="#citing-cocoa-tree" title="Link to this heading">¶</a></h2>
<p>If you use COCOA-Tree in a scientific publication, we would appreciate citations
to the following paper:</p>
<blockquote>
<div><p>BEST PAPER EVER</p>
</div></blockquote>
<p>Bibtex entry:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@article</span> <span class="p">{}</span>
</pre></div>
</div>
</section>
<section id="license-information">
<h2>License information<a class="headerlink" href="#license-information" title="Link to this heading">¶</a></h2>
<p>Conditions on the use and redistribution of this package.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Copyright</span> <span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="mi">2025</span> <span class="n">The</span> <span class="n">COCOA</span><span class="o">-</span><span class="n">Tree</span> <span class="n">developers</span><span class="o">.</span>
<span class="n">All</span> <span class="n">rights</span> <span class="n">reserved</span><span class="o">.</span>

<span class="n">Redistribution</span> <span class="ow">and</span> <span class="n">use</span> <span class="ow">in</span> <span class="n">source</span> <span class="ow">and</span> <span class="n">binary</span> <span class="n">forms</span><span class="p">,</span> <span class="k">with</span> <span class="ow">or</span> <span class="n">without</span>
<span class="n">modification</span><span class="p">,</span> <span class="n">are</span> <span class="n">permitted</span> <span class="n">provided</span> <span class="n">that</span> <span class="n">the</span> <span class="n">following</span> <span class="n">conditions</span> <span class="n">are</span> <span class="n">met</span><span class="p">:</span>

<span class="mf">1.</span> <span class="n">Redistributions</span> <span class="n">of</span> <span class="n">source</span> <span class="n">code</span> <span class="n">must</span> <span class="n">retain</span> <span class="n">the</span> <span class="n">above</span> <span class="n">copyright</span> <span class="n">notice</span><span class="p">,</span>
<span class="n">this</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">conditions</span> <span class="ow">and</span> <span class="n">the</span> <span class="n">following</span> <span class="n">disclaimer</span><span class="o">.</span>

<span class="mf">2.</span> <span class="n">Redistributions</span> <span class="ow">in</span> <span class="n">binary</span> <span class="n">form</span> <span class="n">must</span> <span class="n">reproduce</span> <span class="n">the</span> <span class="n">above</span> <span class="n">copyright</span> <span class="n">notice</span><span class="p">,</span>
<span class="n">this</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">conditions</span> <span class="ow">and</span> <span class="n">the</span> <span class="n">following</span> <span class="n">disclaimer</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">documentation</span>
<span class="ow">and</span><span class="o">/</span><span class="ow">or</span> <span class="n">other</span> <span class="n">materials</span> <span class="n">provided</span> <span class="k">with</span> <span class="n">the</span> <span class="n">distribution</span><span class="o">.</span>

<span class="mf">3.</span> <span class="n">Neither</span> <span class="n">the</span> <span class="n">name</span> <span class="n">of</span> <span class="n">the</span> <span class="n">copyright</span> <span class="n">holder</span> <span class="n">nor</span> <span class="n">the</span> <span class="n">names</span> <span class="n">of</span> <span class="n">its</span> <span class="n">contributors</span>
<span class="n">may</span> <span class="n">be</span> <span class="n">used</span> <span class="n">to</span> <span class="n">endorse</span> <span class="ow">or</span> <span class="n">promote</span> <span class="n">products</span> <span class="n">derived</span> <span class="kn">from</span><span class="w"> </span><span class="nn">this</span> <span class="n">software</span> <span class="n">without</span>
<span class="n">specific</span> <span class="n">prior</span> <span class="n">written</span> <span class="n">permission</span><span class="o">.</span>

<span class="n">THIS</span> <span class="n">SOFTWARE</span> <span class="n">IS</span> <span class="n">PROVIDED</span> <span class="n">BY</span> <span class="n">THE</span> <span class="n">COPYRIGHT</span> <span class="n">HOLDERS</span> <span class="n">AND</span> <span class="n">CONTRIBUTORS</span> <span class="s2">&quot;AS IS&quot;</span>
<span class="n">AND</span> <span class="n">ANY</span> <span class="n">EXPRESS</span> <span class="n">OR</span> <span class="n">IMPLIED</span> <span class="n">WARRANTIES</span><span class="p">,</span> <span class="n">INCLUDING</span><span class="p">,</span> <span class="n">BUT</span> <span class="n">NOT</span> <span class="n">LIMITED</span> <span class="n">TO</span><span class="p">,</span> <span class="n">THE</span>
<span class="n">IMPLIED</span> <span class="n">WARRANTIES</span> <span class="n">OF</span> <span class="n">MERCHANTABILITY</span> <span class="n">AND</span> <span class="n">FITNESS</span> <span class="n">FOR</span> <span class="n">A</span> <span class="n">PARTICULAR</span> <span class="n">PURPOSE</span>
<span class="n">ARE</span> <span class="n">DISCLAIMED</span><span class="o">.</span> <span class="n">IN</span> <span class="n">NO</span> <span class="n">EVENT</span> <span class="n">SHALL</span> <span class="n">THE</span> <span class="n">COPYRIGHT</span> <span class="n">HOLDER</span> <span class="n">OR</span> <span class="n">CONTRIBUTORS</span> <span class="n">BE</span>
<span class="n">LIABLE</span> <span class="n">FOR</span> <span class="n">ANY</span> <span class="n">DIRECT</span><span class="p">,</span> <span class="n">INDIRECT</span><span class="p">,</span> <span class="n">INCIDENTAL</span><span class="p">,</span> <span class="n">SPECIAL</span><span class="p">,</span> <span class="n">EXEMPLARY</span><span class="p">,</span> <span class="n">OR</span>
<span class="n">CONSEQUENTIAL</span> <span class="n">DAMAGES</span> <span class="p">(</span><span class="n">INCLUDING</span><span class="p">,</span> <span class="n">BUT</span> <span class="n">NOT</span> <span class="n">LIMITED</span> <span class="n">TO</span><span class="p">,</span> <span class="n">PROCUREMENT</span> <span class="n">OF</span>
<span class="n">SUBSTITUTE</span> <span class="n">GOODS</span> <span class="n">OR</span> <span class="n">SERVICES</span><span class="p">;</span> <span class="n">LOSS</span> <span class="n">OF</span> <span class="n">USE</span><span class="p">,</span> <span class="n">DATA</span><span class="p">,</span> <span class="n">OR</span> <span class="n">PROFITS</span><span class="p">;</span> <span class="n">OR</span> <span class="n">BUSINESS</span>
<span class="n">INTERRUPTION</span><span class="p">)</span> <span class="n">HOWEVER</span> <span class="n">CAUSED</span> <span class="n">AND</span> <span class="n">ON</span> <span class="n">ANY</span> <span class="n">THEORY</span> <span class="n">OF</span> <span class="n">LIABILITY</span><span class="p">,</span> <span class="n">WHETHER</span> <span class="n">IN</span>
<span class="n">CONTRACT</span><span class="p">,</span> <span class="n">STRICT</span> <span class="n">LIABILITY</span><span class="p">,</span> <span class="n">OR</span> <span class="n">TORT</span> <span class="p">(</span><span class="n">INCLUDING</span> <span class="n">NEGLIGENCE</span> <span class="n">OR</span> <span class="n">OTHERWISE</span><span class="p">)</span>
<span class="n">ARISING</span> <span class="n">IN</span> <span class="n">ANY</span> <span class="n">WAY</span> <span class="n">OUT</span> <span class="n">OF</span> <span class="n">THE</span> <span class="n">USE</span> <span class="n">OF</span> <span class="n">THIS</span> <span class="n">SOFTWARE</span><span class="p">,</span> <span class="n">EVEN</span> <span class="n">IF</span> <span class="n">ADVISED</span> <span class="n">OF</span>
<span class="n">THE</span> <span class="n">POSSIBILITY</span> <span class="n">OF</span> <span class="n">SUCH</span> <span class="n">DAMAGE</span><span class="o">.</span>
</pre></div>
</div>
<div class="docutils container" id="id5">
<div role="list" class="citation-list">
<div class="citation" id="id8" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">DJPV13</a><span class="fn-bracket">]</span></span>
<p>David De Juan, Florencio Pazos, and Alfonso Valencia. Emerging methods in protein co-evolution. <em>Nature Reviews Genetics</em>, 14(4):249–261, April 2013. <a class="reference external" href="https://doi.org/10.1038/nrg3414">doi:10.1038/nrg3414</a>.</p>
</div>
<div class="citation" id="id9" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">eal25</a><span class="fn-bracket">]</span></span>
<p>Margaux et al. Tbd. <em>BioRxiv</em>, 1(1):1, 2025. <a class="reference external" href="https://doi.org/no DOI yet">doi:no DOI yet</a>.</p>
</div>
<div class="citation" id="id6" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id4">HRLR09</a><span class="fn-bracket">]</span></span>
<p>Najeeb Halabi, Olivier Rivoire, Stanislas Leibler, and Rama Ranganathan. Protein Sectors: Evolutionary Units of Three-Dimensional Structure. <em>Cell</em>, 138(4):774–786, August 2009. URL: <a class="reference external" href="https://linkinghub.elsevier.com/retrieve/pii/S0092867409009635">https://linkinghub.elsevier.com/retrieve/pii/S0092867409009635</a> (visited on 2023-07-19), <a class="reference external" href="https://doi.org/10.1016/j.cell.2009.07.038">doi:10.1016/j.cell.2009.07.038</a>.</p>
</div>
<div class="citation" id="id7" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>RRR16<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id1">1</a>,<a role="doc-backlink" href="#id4">2</a>)</span>
<p>Olivier Rivoire, Kimberly A. Reynolds, and Rama Ranganathan. Evolution-Based Functional Decomposition of Proteins. <em>PLOS Computational Biology</em>, 12(6):e1004817, June 2016. <a class="reference external" href="https://doi.org/10.1371/journal.pcbi.1004817">doi:10.1371/journal.pcbi.1004817</a>.</p>
</div>
</div>
</div>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
  </body>
</html>